### DISCUSSION TITLE
Connection problem decpite verfied login credentials: 'Cannot connect to

### DISCUSSION THREAD

**Discussion by TheBaluu** [Hugovdberg/PIconnect#738] [Q&A] [answered]
Hello everybody,

Until now I used Matlab to connect to our PI server via a VPN. In Matlab i can connect to the server without issues using the code:

`server = piclient('<ServerName>', Username="<UserName>", Password="<Password>");   `

Also using the PI SDK I can establish the connection and see the IP address etc.

But now i have to use Python for a project and sadly when I try to conncect using similar code in Python it doesn't connect.

Here is my code:

`import PIconnect as PI`

`print(list(PI.PIServer.servers.keys())) `

`with PI.PIServer(server="<ServerName>", username="<UserName>", password="<Password>") as server:`

`print(server.server_name)`
  

Only by importing the module i get already the following warning:
`C:\Users\HelloThere\AppData\Roaming\Python\Python311\site-packages\PIconnect\PIAF.py:42: InitialisationWarning: Failed loading server data for _<ServerName>_ with error SocketException
  warnings.warn(`

I'm not sure if this is the cause, because even though I get the warning, the following line still executes successfully: `print(list(PI.PIServer.servers.keys()))`

However, the actual connection fails and gives the following error message

`Traceback (most recent call last):`
```
File "C:\Users\HelloThere\Documents\MSc Applied Bioscience\X02 - Preliminary Project\Python_TCPIP\OSIsoft_PI_Server.py", line 10, in <module>
    with PI.PIServer(server="<ServerName>", username="<UserName>", password="<Passwaord>") as server:
```
```
  File "C:\Users\HelloThere\AppData\Roaming\Python\Python311\site-packages\PIconnect\PI.py", line 125, in __enter__
    self.connection.Connect(*self._credentials)
```
`OSIsoft.AF.PI.PIAuthenticationException: **Cannot connect to the PI Data Archive.** `
```
   bei OSIsoft.AF.PI.PIException.ConvertAndThrowException(PIServer piServer, Exception ex, String message)
   bei OSIsoft.AF.PI.PIServer.RemoteConnect(AFConnectionPreference preference, NetworkCredential credential, PIAuthenticationMode authenticationMode)
   bei OSIsoft.AF.PI.PIServer.InternalConnect(Int32 numRetries, IWin32Window owner, AFConnectionPreference preference, NetworkCredential credential, PIAuthenticationMode authenticationMode, Boolean checkConnectingAsSameUser)
   bei OSIsoft.AF.PI.PIServer.Connect(NetworkCredential credential, AFConnectionPreference preference, PIAuthenticationMode authenticationMode)
```

I have seen a similar error in a another discussion here in this forum: https://github.com/Hugovdberg/PIconnect/discussions/622 
However, the difference is that he got the error " Cannot connect to the PI Data Archive. Windows authentication trial failed because the authentication method was not tried. Trust authentication trial failed because insufficient privilege to access the PI Data Archive." which I interpret that my login credentials are correct and my user has "the privilege" to connect. I can replicate this error message when omititng my login credentials and trying to connect using only the server name.

Since it works in Matlab and PI SDK I susspect the problem lies somwhere else. But I am unable to locate the root cause. I dont know if the SocketException raised when importing the module is responsible or if it something else.

In the PI.py file its this piece of code that in the end throws the error but I can not find where the .connection method is to further investigate the problem by myself:

    def __enter__(self):
        if self._credentials:
            self.connection.Connect(*self._credentials)
        else:
            # Don't force to retry connecting if previous attempt failed
            force_connection = False
            self.connection.Connect(force_connection)
        return self

I would be very happy if someone who had the same problem or knows what is going on could help me:)

**Comment by Hugovdberg** [comment, marked as answer]
You mention Windows authentication, so I guess you try to connect with a Windows account, is that right? By default a username/password combination is interpreted as a PI user. For Windows authentication you need to set the authentication mode:

```python
import PIconnect as PI
from PIconnect.PIConsts import AuthenticationMode

with PI.PIServer(username=..., password=..., authentication_mode=AuthenticationMode.WINDOWS_AUTHENTICATION) as server:
    ...
```

Does it work for you using that option?

**Reply by TheBaluu** [reply]
Hey thanks for your quick answer and yes it does work now using the windows AuthenticationMode!👍
Thank you for your help and the PIconnect in general :)  

Btw, you may be interested to know that a warning is still displayed when importing the package: 
`InitialisationWarning: Failed loading server data for <Institution> with error EndpointNotFoundException
  warnings.warn(`
But it still connects and I can retrieve my data so its not really a problem i guess.

**Reply by Hugovdberg** [reply]
You're welcome! This warning is indeed a bit of a struggle. `PIconnect` tries to preload some information on the listed servers, but that might fail due to various reasons. There are three options here: not preloading (might be not too bad after all); raise a warning; or skip the failing servers, but then looking at the list of servers it would appear that the server is not available. In a future version I'm likely to redesign this to only load the server names but not the rest of the metadata.
